driver-core/src/main/java/com/datastax/driver/core/AbstractMonotonicTimestampGenerator.java

```diff
@@ -0,0 +1,86 @@
+/*
+ *      Copyright (C) 2012-2015 DataStax Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.datastax.driver.core;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * Base implementation for monotonic timestamp generators.
+ */
+abstract class AbstractMonotonicTimestampGenerator implements TimestampGenerator {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TimestampGenerator.class);
+
+    private static final long CLOCK_DRIFT_WARNING_INTERVAL = NANOSECONDS.convert(1, SECONDS);
+
+    @VisibleForTesting
+    volatile Clock clock = ClockFactory.newInstance();
+
+    private AtomicLong lastClockDriftWarning = new AtomicLong(Long.MIN_VALUE);
+
+    /**
+     * Compute the next timestamp, given the last timestamp previously generated.
+     * <p/>
+     * The returned timestamp should be strictly greater than the previous one.
+     * If this is not the case, a warning will be logged once every second.
+     * To disable these warnings, set the
+     * {@code com.datastax.driver.core.TimestampGenerator} logger level
+     * to {@code OFF}.
+     * <p/>
+     * This implementation is inspired by {@code org.apache.cassandra.service.ClientState#getTimestamp()}.
+     *
+     * @param last The last timestamp generated by this generator, in microseconds.
+     * @return the next timestamp to use, in microseconds.
+     */
+    protected long computeNext(long last) {
+        long current = clock.currentTimeMicros();
+        if (last >= current) {
+            maybeLogWarning(current, last);
+            return last + 1;
+        }
+        return current;
+    }
+
+    private void maybeLogWarning(long current, long last) {
+        if (LOGGER.isWarnEnabled()) {
+            long now = System.nanoTime();
+            long lastWarning = lastClockDriftWarning.get();
+            if (now > lastWarning + CLOCK_DRIFT_WARNING_INTERVAL && lastClockDriftWarning.compareAndSet(lastWarning, now)) {
+                if (current == last) {
+                    LOGGER.warn(
```

Comment:
making it static does seem like a good idea.  I think this will still happen though, because clock.currentTimeMicros() for `SystemClock` is only going to have millisecond resolution and the likelihood of multiple calls in an ms seems very high.  The previous implementation only printed a warning if 999 entries were created for the same millisecond, so it was much less sensitive to this.